satpy/readers/hdfeos_base.py

```diff
@@ -216,14 +217,48 @@ def load_dataset(self, dataset_name, is_category=False):
         from satpy.readers.hdf4_utils import from_sds
 
         dataset = self._read_dataset_in_file(dataset_name)
-        dask_arr = from_sds(dataset, chunks=CHUNK_SIZE)
+        chunks = self._chunks_for_variable(dataset)
+        dask_arr = from_sds(dataset, chunks=chunks)
         dims = ('y', 'x') if dask_arr.ndim == 2 else None
         data = xr.DataArray(dask_arr, dims=dims,
                             attrs=dataset.attributes())
         data = self._scale_and_mask_data_array(data, is_category=is_category)
 
         return data
 
+    def _chunks_for_variable(self, hdf_dataset):
+        scan_length_250m = 40
+        var_shape = hdf_dataset.info()[2]
+        res_multiplier = self._get_res_multiplier(var_shape)
+        non_yx_chunks = tuple()
+        if len(var_shape) == 3:
+            # assume (band, y, x)
+            non_yx_chunks = ((1,) * var_shape[0],)
+            var_shape = var_shape[-2:]
+        elif len(var_shape) != 2:
+            # don't guess
+            return dask.array.core.normalize_chunks("auto", shape=var_shape, dtype=np.float32)
+        shape_for_250m = tuple(dim_size * res_multiplier for dim_size in var_shape)
+        chunks_for_250m = dask.array.core.normalize_chunks(("auto", -1), shape=shape_for_250m, dtype=np.float32)
+        row_chunks_for_250m = chunks_for_250m[0][0]
+        scanbased_row_chunks_for_250m = np.round(row_chunks_for_250m / scan_length_250m) * scan_length_250m
+        var_row_chunks = scanbased_row_chunks_for_250m / res_multiplier
+        var_row_chunks = max(var_row_chunks, scan_length_250m / res_multiplier)  # avoid getting 0 chunk size
+        return non_yx_chunks + (var_row_chunks, -1)
+
+    @staticmethod
+    def _get_res_multiplier(var_shape):
+        num_columns_to_multiplier = {
+            271: 20,  # 5km
+            1354: 4,  # 1km
+            2708: 2,  # 500m
+            5416: 1,  # 250m
+        }
+        for max_columns, res_multiplier in num_columns_to_multiplier.items():
+            if var_shape[-1] <= max_columns:
+                return res_multiplier
+        return 1
```

Comment:
We already have a mechanism to find out the resolution of a file here: https://github.com/pytroll/satpy/blob/main/satpy/readers/modis_l1b.py#L61-L71

Could we factorise these two?

Reply:
That code gets the resolution for the overall file, but as you know files can contain multiple resolutions of data (geo at one resolution, band data at another, etc). This is another case where if the MODIS readers were completely rewritten and didn't do magic "load by name" functions then we might know this information before loading the variable data.